### PR TITLE
AWS ECS Operator

### PR DESCRIPTION
**Pull request by pedrorjbr**
1. Build a ECS Operator
2. Build a ECS Hook
3. Using "aws_credentials"  as a connection id

### CODE REVIEW

File: airflow/hooks/ecs_hook.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import boto3
+
+class ECSHook(BaseHook):
+    
+    def __init__( self, aws_conn_id="aws_credentials"):
+        self.conn = self.get_connection(aws_conn_id)    
+        self._a_key = None
+        self._s_key = None
+        self.extra_params = json.loads(self.conn.extra)
+        if 'aws_secret_access_key' in self.extra_params:
```
Comment (avram) [inline airflow/hooks/ecs_hook.py]: This prevents boto3 from transparently picking up IAM roles available via the instance's instance profile.

**Comment by pedrorjbr** [comment]
Can someone help me how to test an external service? I am developing an AWS ECS operator.

**Comment by pedrorjbr** [comment]
@mistercrunch Following instructions to move operator and hook to ecs contrib folder. I tested the ECS Operator and already using it, but it is no testable.
How can make it pass to the code coverage?

**Comment by pedrorjbr** [comment]
@mistercrunch: Following what you said in other issue, I moved the ECS Operator and the ECS Hook to the contrib folder. But these two operator arent testable...

**Comment by mistercrunch** [comment]
Can you move the commented tests out of `core.py`? Maybe `airflow/contrib/tests`? We should probably make sure that the `./run_unit_tests.sh` command doesn't find and run those or the CI tests will fail.

**Comment by pedrorjbr** [comment]
@mistercrunch: I reverted the core.py. My integration tests only makes sense on my enviroment so I didnt pushed it anymore. 
I implemented the sftp hook too. I need an operator to copy file from a sftp to S3.

**Comment by pedrorjbr** [comment]
Why am I having erros on code coverage?

File: airflow/contrib/hooks/ecs_hook.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            self._region_key = self.extra_params['region_name']
+        else:
+            raise AirflowException("aws_ecs doesn't exist in the repository")
+        
+    def get_conn(self):
+
+        client = boto3.client('ecs', region_name=self._region_key, 
+                               aws_access_key_id=self._a_key, 
+                               aws_secret_access_key=self._s_key)
+#        client = boto3.client('ecs')
```
Comment (mistercrunch) [inline airflow/contrib/hooks/ecs_hook.py]: remove line

File: airflow/contrib/hooks/ecs_hook.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        return client
+
```
Comment (mistercrunch) [inline airflow/contrib/hooks/ecs_hook.py:27]: rm line

File: airflow/contrib/hooks/sftp_hook.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    
+    def __init__(self,  ftp_conn_id='sftp_default'):
+        self.ftp_conn_id = ftp_conn_id
+        self.conn = None
+        
+    def get_conn(self):
+        """
+        Returns a SFTP connection object
+        """
+        if self.conn is None:
```
Comment (mistercrunch) [inline airflow/contrib/hooks/sftp_hook.py:15]: `if not self.conn:`

File: airflow/contrib/hooks/sftp_hook.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        return self.conn
+    
+    def list_directory(self, path):
+        fileList = self.conn.listdir(path)
+        fileList.reverse()
+        return fileList
+    
+    def get_file(self, from_file, to_file):
+        self.conn.get(from_file, to_file)
+
```
Comment (mistercrunch) [inline airflow/contrib/hooks/sftp_hook.py:28]: remove empty lines

**Comment by mistercrunch** [comment]
I started reading the code and there are leftover comments and the code needs to be linted. Check out the CONTRIBUTING.md and take time to use a linter like `flake8`

File: airflow/contrib/operators/ftp_to_s3.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            os.makedirs(self.tmp_directory)  
+        
+        for file_name  in file_list:
+            s3_key_file = self.s3_key +"/"+str(file_name)
+            exists = s3_hook.check_for_key(s3_key_file, self.s3_bucket)
+            
+            if (exists) and (not self.replace):
+                continue
+            
+            ftp_file_fullpath = self.ftp_folder + "/" + str(file_name)
```
Comment (mistercrunch) [inline airflow/contrib/operators/ftp_to_s3.py]: `os.path.join`

File: airflow/contrib/operators/ftp_to_s3.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            local_file_fullpath = self.tmp_directory + "/" + str(file_name)
+            
+            logging.info("Dowloading file ["+str(ftp_file_fullpath)+"] from sftp to local ["+str(local_file_fullpath)+"]");
+            sftp_hook.get_file(ftp_file_fullpath, local_file_fullpath)
+            logging.info("Done.")
+            logging.info("Uploading file ["+str(local_file_fullpath)+"] to S3 on bucket ["+str(self.s3_bucket)+"] and key ["+str(s3_key_file)+"]")
```
Comment (mistercrunch) [inline airflow/contrib/operators/ftp_to_s3.py]: use `"{}".format()` syntax

File: airflow/contrib/operators/ftp_to_s3.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    :type: s3_conn_id: str
+    :param: filter: a filter function that receiver file_name as parameters.
+    :type: filter: function
+    :param: replace: replace destination file
+    :type: replace: bool
+    :param: tmp_directory: just a temporary directory
+    :type: tmp_directory: str
+    """
+  
+    def __init__(self, s3_bucket, s3_key, ftp_folder, sftp_conn_id='sftp_default', s3_conn_id='s3_default', filter= None, replace = False, tmp_directory = "tmp"):
```
Comment (mistercrunch) [inline airflow/contrib/operators/ftp_to_s3.py]: inconsistent spacing around operator

**Comment by pedrorjbr** [comment]
Thanks! @mistercrunch 

I am sorry for the mistakes, I am kind newbie contributing to open source projects. I will take a look.

**Comment by rfroetscher** [comment]
@pedrorjbr are you still working on this? if not i'd like to take it over

**Comment by pedrorjbr** [comment]
@rfroetscher No, I am not! I made a fork and my team are working with a forked version of airflow. I appreciate any help. Basically we need to put the code in PEP8.

**Comment by vitorbaptista** [comment]
@rfroetscher @pedrorjbr Any updates on this?

**Comment by rfroetscher** [comment]
I'm currently on vacation and will be until 9/12. Feel free to grab this.

**Comment by pedroabi** [comment]
@vitorbaptista: No updates at all.

**Comment by mistercrunch** [comment]
We're cleaning up the open PRs list, please rebase within the next few days or we'll close the PR, in which case you can always reopen the PR from your rebased branch.

**Comment by poulainv** [comment]
+1 !! 🍰

**Pull request closed without merging**